### ISSUE TITLE
Multiple failures in distributed/elastic tests

### ISSUE THREAD

**Issue by Flamefire** [pytorch/pytorch#60260]
## 🐛 Bug

I'm seeing multiple failures in the tests

- distributed/elastic/utils/distributed_test
- distributed/elastic/multiprocessing/api_test

Log shows for the first:
```
 ======================================================================
FAIL: test_port_already_in_use_on_server (__main__.DistributedUtilTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "distributed/elastic/utils/distributed_test.py", line 110, in test_port_already_in_use_on_server
    create_c10d_store(
AssertionError: OSError not raised
```

And for the second:
```
======================================================================
FAIL: test_function_failure_signal (__main__.StartProcessesNotCITest) (start_method='spawn')
run 2x copies of echo3, induce a segfault on first
----------------------------------------------------------------------
Traceback (most recent call last):
  File "distributed/elastic/multiprocessing/api_test.py", line 661, in test_function_failure_signal
    self.assertEqual(os.path.join(log_dir, "0", "error.json"), error_file)
AssertionError: '/tmp/easybuild-tmp/eb-3jgYPy/StartProcess[41 chars]json' != '<N/A>'
- /tmp/easybuild-tmp/eb-3jgYPy/StartProcessesNotCITest_jrdjulk8/tmpuxp713l7/0/error.json
+ <N/A>
```
Which is repeated 8 times with minor variations. Full log starting at that test execution: 
[pterror.txt](https://github.com/pytorch/pytorch/files/6676176/pterror.txt)

## To Reproduce

1. Build and run tests

1. install pytorch
2. checkout tests folder
3. `python distributed/elastic/multiprocessing/api_test.py -v -- StartProcessesNotCITest.test_function_failure_signal`

## Environment

 - PyTorch Version (e.g., 1.0): 1.9.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Python version: 3.8.6

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

**Comment by Flamefire** [comment]
I'm really wondering how those tests were ever accepted into the PyTorch codebase.
The `test_function_failure_signal` cannot work at all. It (purposely) creates a SIGSEGV which terminates the subprocess while NOT creating an exception and specifically NOT a `ChildFailedError` as would be required for the test to work. After some debugging and digging into that mess I discovered that those tests are specifically excluded from PyTorchs CI.

This makes me wonder: If those tests can basically not be run on Windows, OSX, TSAN, ASAN and PyTorchs own CI, what do they add to the code quality? Wouldn't it be better then to remove them to avoid such situations as this?

**Comment by Flamefire** [comment]
And the other failing test `test_port_already_in_use_on_server` (added by @kiukchung) fails because on Linux either TCPv4 or TCPv6 can be used. The test tries to block a port by creating a socket using an unspecified protocol. The TCPStore does the same so what happens (in my case at least) is that the Python socket is using TCPv6 and the TCPStore uses TCPv4

**Comment by cbalioglu** [comment]
Those are tests that are moved from the standalone TorchElastic repo as part of the upstreaming work we have done in PyTorch v1.9. Both authors (Aliaksandr and Kiuk) as well as the PyTorch Distributed team are aware of those local run failures. We already have an internal backlog item (T90265589) to fix them in the near future.

**Comment by rohan-varma** [comment]
> Those are tests that are moved from the standalone TorchElastic repo as part of the upstreaming work we have done in PyTorch v1.9. Both authors (Aliaksandr and Kiuk) as well as the PyTorch Distributed team are aware of those local run failures. We already have an internal backlog item (T90265589) to fix them in the near future.

@cbalioglu Should we remove these tests if they aren't intended to/don't work in OSS CI? It looks like one of the tests is always skipped in PyTorch CI. If it's only meant to run internally, can we check it into an internal-only file? Ideally all tests in OSS should always run in OSS under some CI config.

**Comment by kiukchung** [comment]
@tierex is going to move `test_function_failure_signal` to an fb-only directory. Reason it fails in OSS is because we don't register signal handlers that produce detailed stack traces on signals, which we do by default for FB builds.

**Comment by Flamefire** [comment]
Thanks for the reply. I wouldn't even mind them having the tests here even if they are meant to be run internally only. But instead of an opt-in to skip them they should have an opt-in to run them if they only work in very specific environments.

Just so you understand my displeasure: We have to build PyTorch from source for our HPC users so they can e.g. use it on A100 GPUs, POWER or ARM architectures or with the vendor provided environment (CUDA drivers and such). To verify that this worked the tests are essential and differentiating between faulty tests and faulty code (every PyTorch release so far adds both) is very hard and takes a considerable amount of time which I don't really have.
Of course this isn't unique to PyTorch, TensorFlow has the same issues although there usually the tests are fine but the build fails or the code has actual bugs (well, that isn't exactly better...)

@kiukchung What about the IPv4/IPv6 issue of test_port_already_in_use_on_server? I guess you could create 2 TCP stores on the same port or open another socket first to really trigger this error in the test. But as of now this test is basically guaranteed to fail if IPv4 & IPv6 are both available.

**Comment by kiukchung** [comment]
@Flamefire thanks for the background. I completely understand that false negative tests can be frustrating. In this case the offending test was a result of the subtle differences between fb-internal and oss build/runtime environment. We are fully aware of this broader issue and are working to close the gap so that the PyTorch UX and DevX is standardized. In the meanwhile thank you for your patience.

As for the IPv4/IPv6 issue - this points to a known issue we have in TCPStore where, precisely to your point, when the host is dual-stacked but is missing an ipv4 DNS and/or preferred ordering specified in `/etc/gai.conf` the listen and connect sockets can bind onto IPv4 vs IPv6. The fact that this test is flaky probably means that if you try running a single host, multi process distributed torch chances are it will fail (with a TimeoutError) since the TCPStore (on rank 0) will eventually timeout waiting for all workers to connect. We'd have to do either of these two things to resolve this flaky test:

1.  (Preferred) Merge this PR that fixes TCPStore: https://github.com/pytorch/pytorch/pull/49124 - I'll circle back with the author to see where we are in terms of committing it.

-- OR --

2. (Easy short term) In `torch.distributed.elastic.utils.distributed.get_socket_with_port()`, check for whether the host is dual-stack and prefer to bind to IPv4 if so.

**Comment by Flamefire** [comment]
> The fact that this test is flaky probably means that if you try running a single host, multi process distributed torch chances are it will fail (with a TimeoutError) 

I think you are referring to another test than I did. I was talking about `test_port_already_in_use_on_server` which is not flaky but guaranteed to fail if the host is dual-stacked. The linked PR likely does NOT fix the failure according to the title "Default to IPv6 on listen" because on a dual-stacked host the TCPStore will try the IPv6 port first and (if that was taken by the Python Socket setup by the test which in turn may use user v4 or v6) fall  back to IPv4 in the second iteration of this loop: https://github.com/pytorch/pytorch/blob/3846cef2d7e3f902d13db64a9147ce7a35cb2c8e/torch/lib/c10d/Utils.cpp#L170-L193

So basically the assumption of the test, that "the port is already in use" is broken on dual-stacked machines.

**Comment by kiukchung** [comment]
I've fixed the offending test in https://github.com/pytorch/pytorch/pull/60558 by using `TCPStore` itself to "reserve" a port. Creating two TCPStores on the same port should make binding to an ipv4 vs ipv6 port deterministic within the same host with the same network settings.

**Comment by Flamefire** [comment]
I'm not exactly sure why this works. I expected this to fail due to the reason explained in my previous comment.

I ran the new test code on top of PyTorch 1.9 and it indeed failed (i.e. test succeeded) but I'm not sure why. After creating the store1 I called `socket.getaddrinfo(None, store1.port, family=socket.AF_UNSPEC, type=socket.SOCK_STREAM, flags=socket.AI_PASSIVE|socket.AI_ADDRCONFIG)` which is the Python equivalent of that the TCPStore does. It returns `[(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('0.0.0.0', 34682)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::', 34682, 0, 0))]`, i.e. the IPv4 and IPv6 entries for that port.

And according to the above code both of those should be tried and as (I assume) TCPStore only listens on one of them the other should be free. So what am I missing?

Side note: `setSocketAddrReUse` is called which on Windows disables SO_REUSEADDR but on anything else enables it. Not sure if this is intentional.

**Comment by Flamefire** [comment]
Ok I did a couple more experiments, even reverting the order in which the TCPStore tries the addrinfo entires:

- When using a Python socket (tested many variations) the Python socket will bind via IPv6. When TCPStore tries that too, it fails but then succeeds with IPv4
- When using a TCPStore to block the socket then both IPv4 and IPv6 are blocked. I haven't found why. Maybe Python does something to keep the other protocol untouched?

**Comment by kiukchung** [comment]
Thanks for running the experiments! So your outstanding question is why `store2` does not fall back to binding and listening to an IPv6 socket after `store1` is listening on an IPv4 socket.

It is indeed strange. According to the ipv6 man pages (https://man7.org/linux/man-pages/man7/ipv6.7.html):

```
IPv4 and IPv6 share the local port space.  When you get an IPv4
connection or packet to an IPv6 socket, its source address will
be mapped to v6 and it will be mapped to v6.
```

and also
```
IPV6_V6ONLY (since Linux 2.4.21 and 2.6)
              If this flag is set to true (nonzero), then the socket is
              restricted to sending and receiving IPv6 packets only.  In
              this case, an IPv4 and an IPv6 application can bind to a
              single port at the same time.

              If this flag is set to false (zero), then the socket can
              be used to send and receive packets to and from an IPv6
              address or an IPv4-mapped IPv6 address.

              The argument is a pointer to a boolean value in an
              integer.

              The default value for this flag is defined by the contents
              of the file /proc/sys/net/ipv6/bindv6only.  The default
              value for that file is 0 (false).
```

I don't see any use of `IPV6_V6ONLY` in code (in python nor in TCPStore, nor anywhere in pytorch). So unless your system settings have it turned on (in `/proc/sys/net/ipv6/bindv6only`), then the results of your second experiment is expected (e.g. `store1` binds to port `n` and the second time around you shouldn't be able to bind on that port in either ipv4 or ipv6).

My wild guess is that something is making sockets created from python implicity set `IPV6_V6ONLY` can you try to see what the socket options are for the python socket you created in your first experiment using `socket.getsockopt()` - https://docs.python.org/3/library/socket.html#socket.socket.getsockopt?

**Comment by Flamefire** [comment]
> My wild guess is that something is making sockets created from python implicity set IPV6_V6ONLY

You were on spot:
```
>>> import socket
>>> addrs = socket.getaddrinfo(host="localhost", port=None, family=socket.AF_UNSPEC, type=socket.SOCK_STREAM)
>>> family, type, proto, _, _ = addrs[0]
>>> s = socket.socket(family, type, proto)
>>> s.getsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY)
0
>>> s.bind(("localhost", 0))
>>> s.getsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY)
1
```

I have no idea why and can't find anything in the cpython sources why that is the case. /proc/sys/net/ipv6/bindv6only is also 0

Anyway this actually uncovers a new potential failure point: If /proc/sys/net/ipv6/bindv6only was set to 1 then the new test would fail, wouldn't it?

--> use the python socket but explicitly set IPV6_V6ONLY=0 (ignoring errors when that doesn't exist) or (probably good idea anyway) set it explicitly (where available) in the TCPStore

**Comment by kiukchung** [comment]
Yes you are correct that if a system setting that triggers bindv6only then the new test will also fail. For this particular edgecase I'd like to document that the expectation is that you don't set the bindv6only for the build host and rank 0. The reason is that you probably want to bind and be able to serve both ipv6 and ipv4 connections from the tcpstore server. Let me know what you think.

**Comment by Flamefire** [comment]
I wouldn't rely on a system setting here. Why not set it explicitly (maybe "just" by default)?

At least the tests should be resilient against system settings. See my proposal to explicitly block both port types.

**Comment by kiukchung** [comment]
Network system settings are tricky to guard against purely from the software. While largely I do agree with you that software should be system-independent, in this particular case the incompatibility with v6only setting extends beyond the success/failure of the test. In its current implementation, TCPStore itself is not really compatible with implicit v6only setting (especially in the case where the server has it turned on but the workers do not and the dns routes with ipv4).

If you have a particular use-case for having to set v6only as system (not socket) settings, feel free to submit a PR with the proposed changes and I'd be happy to review it. Thanks!

**Comment by Flamefire** [comment]
Why is this particular setting tricky to guard against? It is literally a single function call. I have proposed 2 alternatives so at least the test is system independent (and IMO tests must be as independent as possible as false positives are a problem)

> --> use the python socket but explicitly set IPV6_V6ONLY=0 (ignoring errors when that doesn't exist) or (probably good idea anyway) set it explicitly (where available) in the TCPStore

And:

>  TCPStore itself is not really compatible with implicit v6only setting

Erm. So why not make it explicit not-v6only? And possibly provide an opt-out where this isn't wanted?

>  having to set v6only as system (not socket) settings

I'm just proposing to (un)set this for the socket(s). At the very least for the test. Is there any downside I'm missing?
